addons/fleet/fleet.py

```diff
@@ -1,818 +1,725 @@
-        #need to cast in float because the value receveid from web client maybe an integer (Javascript and JSON do not
-        #make any difference between 3.0 and 3). This cause a problem if you encode, for example, 2 liters at 1.5 per
-        #liter => total is computed as 3.0, then trigger an onchange that recomputes price_per_liter as 3/2=1 (instead
-        #of 3.0/2=1.5)
-        #If there is no change in the result, we return an empty dict to prevent an infinite loop due to the 3 intertwine
-        #onchange. And in order to verify that there is no change in the result, we have to limit the precision of the 
-        #computation to 2 decimal
-        liter = float(liter)
-        price_per_liter = float(price_per_liter)
-        amount = float(amount)
-        if amount > 0 and liter > 0 and round(amount/liter,2) != price_per_liter:
-            return {'value': {'price_per_liter': round(amount / liter,2),}}
-        elif amount > 0 and price_per_liter > 0 and round(amount/price_per_liter,2) != liter:
-            return {'value': {'liter': round(amount / price_per_liter,2),}}
-        elif liter > 0 and price_per_liter > 0 and round(liter*price_per_liter,2) != amount:
-            return {'value': {'amount': round(liter * price_per_liter,2),}}
-        else :
-            return {}
-
-    def _get_default_service_type(self, cr, uid, context):
+        liter = float(self.liter)
+        price_per_liter = float(self.price_per_liter)
+        amount = float(self.amount)
+        if liter > 0 and price_per_liter > 0 and round(liter * price_per_liter, 2) != amount:
+            self.amount = round(liter * price_per_liter, 2)
+        elif amount > 0 and liter > 0 and round(amount / liter, 2) != price_per_liter:
+            self.price_per_liter = round(amount / liter, 2)
+        elif amount > 0 and price_per_liter > 0 and round(amount / price_per_liter, 2) != liter:
+            self.liter = round(amount / price_per_liter, 2)
+
+    @api.onchange('price_per_liter')
+    def on_change_price_per_liter(self):
+        """
+        need to cast in float because the value received from web client maybe an integer (Javascript and JSON do not
+        make any difference between 3.0 and 3). This cause a problem if you encode, for example, 2 liters at 1.5 per
+        liter => total is computed as 3.0, then trigger an onchange that recomputes price_per_liter as 3/2=1 (instead
+        of 3.0/2=1.5)
+        If there is no change in the result, we return an empty dict to prevent an infinite loop due to the 3 intertwine
+        onchange. And in order to verify that there is no change in the result, we have to limit the precision of the
+        computation to 2 decimal
+        """
+        liter = float(self.liter)
+        price_per_liter = float(self.price_per_liter)
+        amount = float(self.amount)
+        if liter > 0 and price_per_liter > 0 and round(liter * price_per_liter, 2) != amount:
```

Comment:
There is no need to cast those values to `float`, since the cast operation has been done when the values have been put in cache. So `self.liter`, `self.price_per_liter` and `self.amount ` are guaranteed to be floats.